### PR TITLE
Test DualView resize/realloc with WithoutInitializing for types without default constructor

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request adds another test for `DualView` for https://github.com/kokkos/kokkos/pull/6993 that fixed `Kokkos::realloc/resize` with Kokkos::WithouttInitializing such that it can be used with types that don't define a default constructor.

### CODE REVIEW

File: containers/unit_tests/TestDualView.hpp [hunk 1]
```diff
@@ -463,12 +472,36 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
...
+struct Kokkos::reduction_identity<Test::NoDefaultConstructor> {
+  static Test::NoDefaultConstructor sum() { return {0}; }
+};
+
+namespace Test {
+
 TEST(TEST_CATEGORY, dualview_realloc) {
   test_dualview_realloc<int, TEST_EXECSPACE>();
+  Impl::test_dualview_realloc<NoDefaultConstructor, TEST_EXECSPACE,
+                              /* Initialize */ false>();
```
Comment (cedricchevalier19) [inline containers/unit_tests/TestDualView.hpp:498]: A little bit disturbing to have `Impl::` on this line and not on the one above.

File: containers/unit_tests/TestDualView.hpp [hunk 2]
```diff
@@ -463,12 +472,36 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
...
 }
 
 TEST(TEST_CATEGORY, dualview_resize) {
   test_dualview_resize<int, TEST_EXECSPACE>();
+  Impl::test_dualview_resize<NoDefaultConstructor, TEST_EXECSPACE,
+                             /* Initialize */ false>();
```
Comment (cedricchevalier19) [inline containers/unit_tests/TestDualView.hpp:504]: A little bit disturbing to have `Impl::` on this line and not on the one above.

**Review by cedricchevalier19 — APPROVED** [review]

File: containers/unit_tests/TestDualView.hpp (hunk 1 above)
Comment (masterleinad) [reply containers/unit_tests/TestDualView.hpp:498]: `test_dualview_realloc` calls `Impl::test_dualview_realloc` both with `Initialiize=true` and `Initialize=false`. We can only test with `Initialize=false` for `NoDefualtConstructor`, though.

**Review by crtrott — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
This doesn't work yet the failures in Jenkins are real.

**Comment by dalg24** [comment]
Retest this please

**Review by crtrott — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
There is a SYCL segfault is that real?

**Comment by masterleinad** [comment]
> There is a SYCL segfault is that real?

No,
> [ RUN      ] sycl_graph.launch_six
 3/57 Test  #3: Kokkos_CoreUnitTest_SYCL1A .......................***Exception: SegFault 30.80 sec

was fixed in a different pull request and this one only modifies container tests anyway.

**Review by crtrott — APPROVED** [review]

**Pull request merged**